crates/sui-transactional-test-runner/src/test_adapter.rs

```diff
@@ -589,11 +600,153 @@ impl<'a> MoveTestAdapter<'a> for SuiTestAdapter<'a> {
+                        self.compiled_state()
+                            .add_and_generate_interface_file(module);
+                    }
+                };
+                Ok(merge_output(warnings_opt, output))
+            }
         }
     }
 }
 
+fn merge_output(left: Option<String>, right: Option<String>) -> Option<String> {
+    match (left, right) {
+        (None, right) => right,
+        (left, None) => left,
+        (Some(mut left), Some(right)) => {
+            left.push_str(&right);
+            Some(left)
+        }
+    }
+}
+
+macro_rules! move_call {
+    {$builder:expr, ($addr:expr)::$module_name:ident::$func:ident($($args:expr),* $(,)?)} => {
+        $builder.programmable_move_call(
+            $addr,
+            ident_str!(stringify!($module_name)).to_owned(),
+            ident_str!(stringify!($func)).to_owned(),
+            vec![],
+            vec![$($args),*],
+        )
+    }
+}
+
 impl<'a> SuiTestAdapter<'a> {
+    fn upgrade_package(
+        &mut self,
+        current_package_id: Option<String>,
+        modules: Vec<(Option<Symbol>, CompiledModule)>,
+        upgrade_capability: u64,
+        gas_budget: Option<u64>,
+    ) -> anyhow::Result<(Option<String>, Vec<(Option<Symbol>, CompiledModule)>)> {
+        let current_package_id = current_package_id
+            .map_or(ObjectID::from_address(AccountAddress::ZERO), |id| {
+                ObjectID::from_str(&id).unwrap()
+            });
+        let all_bytes = modules
+            .iter()
+            .map(|(_, module)| {
+                let mut module_bytes = vec![];
+                module.serialize(&mut module_bytes)?;
+                Ok(module_bytes)
+            })
+            .collect::<anyhow::Result<Vec<Vec<u8>>>>()?;
+        let gas_budget = gas_budget.unwrap_or(DEFAULT_GAS_BUDGET);
+        let dependencies = system_package_ids();
+        let mut builder = ProgrammableTransactionBuilder::new();
+
+        let _cap = SuiValue::Object(upgrade_capability).into_argument(&mut builder, self)?;
+
+        let upgrade_arg = builder.pure(UpgradePolicy::COMPATIBLE).unwrap();
+        let digest = TransactionDigest::new(self.rng.gen());
+        let digest_arg = builder.pure(digest).unwrap();
+
+        let upgrade_ticket = move_call! {
+            builder,
+            (SuiFramework::ID)::package::authorize_upgrade(Argument::Input(0), upgrade_arg, digest_arg)
+        };
+
+        let upgrade_receipt =
+            builder.upgrade(current_package_id, upgrade_ticket, dependencies, all_bytes);
+
+        move_call! {
+            builder,
+            (SuiFramework::ID)::package::commit_upgrade(Argument::Input(0), upgrade_receipt)
+        };
+
+        let pt = builder.finish();
+
+        let data = |sender, gas| {
+            TransactionData::new_programmable_with_dummy_gas_price(
+                sender,
+                vec![gas],
+                pt,
+                gas_budget,
+            )
+        };
```

Comment:
~Roughly based off of https://github.com/MystenLabs/sui/blob/97b411ab1824e1383d3a68b36587e2e86eaf4a8b/crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs#L164-L184